bika/lims/utils/analysisrequest.py

```diff
@@ -549,3 +469,99 @@ def resolve_rejection_reasons(values):
         return [{"selected": selected, "other": other}]
 
     return []
+
+
+def do_rejection(sample, notify=None):
+    """Rejects the sample and if succeeds, generates the rejection pdf and
+    sends a notification email. If notify is None, the notification email will
+    only be sent if the setting in Setup is enabled
+    """
+    sample_id = api.get_id(sample)
+    if not sample.getRejectionReasons():
+        logger.warn("Cannot reject {} w/o rejection reasons".format(sample_id))
+        return
+
+    success, msg = doActionFor(sample, "reject")
+    if not success:
+        logger.warn("Cannot reject the sample {}".format(sample_id))
+        return
+
+    # Generate a pdf with the rejection reasons
+    pdf = get_rejection_pdf(sample)
+
+    # Attach the PDF to the sample
+    filename = "{}-rejected.pdf".format(sample_id)
+    sample.createAttachment(pdf, filename=filename)
+
+    # Do we need to send a notification email?
+    if notify is None:
+        setup = api.get_setup()
+        notify = setup.getNotifyOnSampleRejection()
+
+    if notify:
+        # Compose the email
+        mime_msg = get_rejection_mail(sample)
+        # Attach the pdf
+        mime_msg.attach(to_email_attachment(pdf))
+        # Send the email
+        send_email(mime_msg)
+
+
+def get_rejection_pdf(sample):
+    """Generates a pdf with sample rejection reasons
+    """
+    # Avoid circular dependencies
+    from bika.lims.browser.analysisrequest.reject import \
+        AnalysisRequestRejectPdfView
+
+    # Render the html's rejection document
+    tpl = AnalysisRequestRejectPdfView(sample, api.get_request())
+    html = tpl.template()
+    html = safe_unicode(html).encode("utf-8")
+
+    # Generate the pdf
+    return createPdf(htmlreport=html)
+
+
+def get_rejection_mail(sample):
+    """Generates an email to sample contacts with rejection reasons
+    """
+    # Avoid circular dependencies
+    from bika.lims.browser.analysisrequest.reject import \
+        AnalysisRequestRejectEmailView
+
+    # Render the email body
+    tpl = AnalysisRequestRejectEmailView(sample, api.get_request())
+    email_body = tpl.template()
+
+    def to_valid_email_address(contact):
+        if not contact:
+            return None
+        address = contact.getEmailAddress()
+        if not is_valid_email_address(address):
+            return None
+        name = contact.getFullname() or ""
+        return to_email_address(address, name=name)
```

Comment:
I would rather rely on the email only and skip the name part. I can remember that I encountered some trouble with the email server (Postfix) when I used it.  Just pass the email only like described in that doctest: https://github.com/senaite/senaite.core/blob/c1f98444f0d33cb5d8f3e12c104c13b4c5ce732a/bika/lims/tests/doctests/API_mail.rst#email-address

Reply:
yep, done